src/workerd/util/sqlite.c++

```diff
@@ -466,21 +466,123 @@ kj::StringPtr SqliteDatabase::getCurrentQueryForDebug() {
+        // all savepoints implicitly.
+        savepoints.clear();
+        inTransaction = false;
+      }
+
+      if (savepoints.empty() && !inTransaction) {
+        // Transaction stack is empty, so the transaction is committed. We can release the rollback
+        // callbacks.
+        rollbackCallbacks.clear();
+      }
+    }
+
+    KJ_CASE_ONEOF(rollback, RollbackTxn) {
+      KJ_IF_SOME(name, rollback.savepointName) {
+        for (;;) {
+          KJ_ASSERT(!savepoints.empty(), "released a savepoint that didn't exist?");
+          if (savepoints.back().name == name) {
+            // Found the saveponit.
+            // Call all rollback callbacks later than the savepoint.
+            size_t index = savepoints.back().rollbackCallbackIndex;
+            KJ_ASSERT(rollbackCallbacks.size() >= index);
+            while (rollbackCallbacks.size() > index) {
+              rollbackCallbacks.back()();
+              rollbackCallbacks.removeLast();
+            }
+
+            // NOTE: Rolling back to a savepoint does not actually release the savepoint. Hence
+            //   we seave this savepoint as the last item in `savepoints`. It must be released
+            //   separately.
+            break;
+          }
+
+          savepoints.removeLast();
+        }
+      } else {
+        KJ_ASSERT(inTransaction, "ROLLBACK TRANSACTION without BEGIN TRANSACTION?");
+
+        savepoints.clear();
+        inTransaction = false;
+
+        while (!rollbackCallbacks.empty()) {
+          rollbackCallbacks.back()();
+          rollbackCallbacks.removeLast();
+        }
+      }
+    }
+  }
+}
+
 // Set up the regulator that will be used for authorizer callbacks while preparing this
 // statement.
-kj::Own<sqlite3_stmt> SqliteDatabase::prepareSql(
+SqliteDatabase::StatementAndEffect SqliteDatabase::prepareSql(
     const Regulator& regulator, kj::StringPtr sqlCode, uint prepFlags, Multi multi) {
   sqlite3* db = &KJ_ASSERT_NONNULL(maybeDb, "previous reset() failed");
 
-  KJ_ASSERT(currentRegulator == kj::none, "recursive prepareSql()?");
+  ParseContext parseContext;
+  KJ_ASSERT(currentParseContext == kj::none, "recursive prepareSql()?");
+  KJ_DEFER(currentParseContext = kj::none);
+  currentParseContext = parseContext;
+
+  KJ_ASSERT(currentRegulator == kj::none,
+      "can't prepare statements inside executeWithRegulator() callback");
   KJ_DEFER(currentRegulator = kj::none);
   currentRegulator = regulator;
 
   for (;;) {
     sqlite3_stmt* result;
     const char* tail;
 
-    SQLITE_CALL(sqlite3_prepare_v3(db, sqlCode.begin(), sqlCode.size(), prepFlags, &result, &tail));
+    auto prepareResult =
+        sqlite3_prepare_v3(db, sqlCode.begin(), sqlCode.size(), prepFlags, &result, &tail);
+    if (prepareResult == SQLITE_AUTH) {
+      KJ_IF_SOME(error, parseContext.authError) {
+        // Throw the tailored auth error.
+        kj::throwFatalException(kj::mv(error));
+      }
+    }
+    if (prepareResult != SQLITE_OK) {
```

Comment:
```suggestion
    } else if (prepareResult != SQLITE_OK) {
```

Reply:
No, that would introduce a bug in the case that `parseContext.authError` is not set.

Reply:
A quick comment here might be useful but good either way.